### PR TITLE
Add types to Checkout State Provider, and introduce Context Module Functions

### PR DESCRIPTION
**Pull request by mikejolley**
This PR implements Typescript in the Checkout State Provider. Doing this revealed a bunch of typos and bugs! 

As well as this, I added some shared helpers to base-utils to help with Typescript support (isObject, objectHasProp, isNull).

Finally, inspired by epic react I added some new functions exported from the context provider which take `dispatch` as a parameter and do something. These were previously exported by the context provider, but with this method we only need to export the dispatcher which means the functions can be tree-shaken.

- `submitCheckout`
- `resetCheckout`
- `setRedirectUrl`
- `setHasError`
- `incrementCalculating`
- `decrementCalculating`
- `setCustomerId`
- `setOrderId`
- `setOrderNotes`
- `setShouldCreateAccount`
- `setAfterProcessing`

So example usage would be:

```javascript
import {
	useCheckoutContext,
	setOrderNotes,
} from '@woocommerce/base-context/cart-checkout/checkout-state';

...

const { dispatch } = useCheckoutContext();

setOrderNotes( dispatch, value );
```

**To test:**

Smoke test the checkout process, including making payment. Checkout state provider is only used during checkout.

### CODE REVIEW

File: assets/js/base/context/cart-checkout/checkout-state/types.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	customerId: number;
+	shouldCreateAccount: boolean;
+	processingResponse: PaymentResultDataType | null;
+};
+
+export type CheckoutStateContextType = {
+	// Checkout state provider dispatch function.
+	dispatch: React.Dispatch< ActionType >;
+	// Submits the checkout and begins processing.
+	onSubmit: () => void;
```
Comment (mikejolley) [inline assets/js/base/context/cart-checkout/checkout-state/types.ts:53]: I've left this in place because it's exposed to Payment Methods.

Comment (nerrad) [reply assets/js/base/context/cart-checkout/checkout-state/types.ts:53]: Potentially, the `usePaymentMethodIntegrations` hook _could_ just pull in the dispatch functions and expose them though right?

**Comment by mikejolley** [comment]
I am extracting the TS changes to a new PR.

**Comment by mikejolley** [comment]
Closing in favour of https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4200 and https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4204

**Pull request closed without merging**